java-checks/src/main/java/org/sonar/java/checks/ThreadLocalCleanupCheck.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5164")
+public class ThreadLocalCleanupCheck extends IssuableSubscriptionVisitor {
+
+  private static final String THREAD_LOCAL = "java.lang.ThreadLocal";
+  private static final MethodMatcher THREADLOCAL_SET = MethodMatcher.create()
+    .typeDefinition(THREAD_LOCAL).name("set").addParameter(TypeCriteria.anyType());
+  private static final MethodMatcher THREADLOCAL_REMOVE = MethodMatcher.create()
+    .typeDefinition(THREAD_LOCAL).name("remove").withoutParameter();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      Symbol.TypeSymbol clazz = ((ClassTree) tree).symbol();
+      clazz.memberSymbols().stream()
+        .filter(Symbol::isVariableSymbol)
+        .filter(s -> s.type().is(THREAD_LOCAL))
+        .filter(ThreadLocalCleanupCheck::isNotSuper)
+        .forEach(this::checkThreadLocalField);
+    } else {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      if (THREADLOCAL_SET.matches(mit) && mit.arguments().get(0).is(Tree.Kind.NULL_LITERAL)) {
+        reportIssue(mit, "Use \"remove()\" instead of \"set(null)\".");
```

Comment:
I don't see why, in the description, this would not work. If it's a common mistake, I would add a small text to make it explicit.

Reply:
I added some explanation to the RSPEC